django/db/models/fields/related.py

```diff
@@ -298,10 +281,10 @@ def contribute_to_class(self, cls, name, virtual_only=False):
             self.rel.related_name = related_name
         other = self.rel.to
         if isinstance(other, six.string_types) or other._meta.pk is None:
-            def resolve_related_class(field, model, cls):
+            def resolve_related_class(cls, model, field):
                 field.rel.to = model
                 field.do_related_class(model, cls)
-            add_lazy_relation(cls, self, other, resolve_related_class)
+            lazy_related_operation(resolve_related_class, cls, other, field=self)
```

Comment:
I've simplified ``create_many_to_many_intermediary_model`` by eliminating these kinds of checks, but in this particular spot, removing the ``other._meta.pk`` check causes a schema editor test to fail.

It's related to the way the sqlite backend performs column deletion:
* create a table with the updated definition called "app_model_new"
* copy the data
* delete the "app_model" table
* rename "app_model_new" to "app_model"

Consider the situation of a recursive relationship. In the old code, ``field.rel.to`` always points at the original mode with table name "app_model" table. If you remove the check so the callback always runs, ``field.rel.to`` ends up pointing at a model with table name "app_model_new". After you rename the table back to "app_model", its definition is left with a dangling reference to "app_model_new".

An equivalent way to do this should be:
* rename "app_model" to "app_model_new"
* create a table with the updated definition called "app_model"
* copy the data
* delete the "app_model_new" table

To me this seems a little more straightforward – there's [a bit of awkward dealing with the "_new" table name in the schema editor](https://github.com/django/django/blob/19188826b4aa989475418f1ea9bf8631b04da1e8/django/db/backends/sqlite3/schema.py#L152) which could be avoided.

Happy to tackle this change if you agree it's sensible.